inlanecruising_plugin/src/object_avoidance.cpp

```diff
@@ -0,0 +1,189 @@
+
+#include <inlanecruising_plugin/object_avoidance.h>
+
+namespace inlanecruising_plugin
+{
+namespace object_avoidance
+{
+    ObjectAvoidance::ObjectAvoidance(){}
+
+    cav_msgs::TrajectoryPlan ObjectAvoidance::update_traj_for_object(cav_msgs::TrajectoryPlan& original_tp, const carma_wm::WorldModelConstPtr& wm_, double current_speed_) 
+    {
+        
+        cav_msgs::TrajectoryPlan update_tpp_vector;
+
+        geometry_msgs::Twist current_velocity;
+        current_velocity.linear.x = current_speed_;
+
+        // // TODO get roadway object
+        std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+        cav_msgs::RoadwayObstacleList rwol2;
+        rwol2.roadway_obstacles = rwol;
+
+        std::cout << "host_vehicle_size" << host_vehicle_size.x << std::endl;
+
+        std::vector<cav_msgs::RoadwayObstacle> rwol_collision = carma_wm::collision_detection::WorldCollisionDetection(rwol2, original_tp, host_vehicle_size, current_velocity, 10.0);// 10 replace to target_time
+        std::cout << "rwol_collision" << rwol_collision.size() << std::endl;
+        std::cout << "rwol" << rwol.size() << std::endl;
+
+        // correct the input types
+        if(rwol_collision.size() > 0) {
+
+            std::cout << "Collision!" << std::endl;
+
+            // use trajectory utiles to update trajectory plan points
+            
+            // lead vehicle trjactory
+            double dist_x = rwol_collision[0].object.pose.pose.position.x - original_tp.trajectory_points[0].x;
+            double dist_y = rwol_collision[0].object.pose.pose.position.y - original_tp.trajectory_points[0].y;
+            double x_lead = sqrt(dist_x*dist_x + dist_y*dist_y);
+
+            // roadway object position
+            double gap_time = (x_lead - x_gap)/current_speed_;
+
+            double collision_time = 0; //\TODO comming from carma_wm collision detection in future (not used now)
+
+            double goal_velocity = rwol_collision[0].object.velocity.twist.linear.x;
+
+            double goal_pos = x_lead - goal_velocity * gap_time - 10.0; 
+
+            double initial_time = 0;
+            double initial_pos = 0.0; //relative initial position (first trajectory point)
+
+            double initial_accel = 0;
+            double goal_accel = 0;
+
+            // TODO: Add special case for when the object is not moving
+
+            double delta_v_max = abs(rwol_collision[0].object.velocity.twist.linear.x - max_trajectory_speed(original_tp.trajectory_points));
+            
+            double t_ref = (original_tp.trajectory_points[original_tp.trajectory_points.size() - 1].target_time.toSec() - original_tp.trajectory_points[0].target_time.toSec());
+
+            double t_ph = 4 * delta_v_max / maximum_deceleration_value;
+
+            double tp = 0;
+
+            if(t_ph > tpmin && t_ref < t_ph){
```

Comment:
Is this supposed to be `... && t_ph < t_ref)`? Looks like you're trying to clamp tp to be between `tpmin` (lower bound) and `t_ref` (upper bound), but I think this inequality is inverted for that.

Reply:
updated